doc/decisions/DR016-Related-entities-result-paging.md

```diff
@@ -0,0 +1,333 @@
+
+Paging through an entire batch via a cursor object was rejected mainly
+due to being broadly similar to a whole-batch page token. A whole-batch
+cursor object offers few additional benefits, whilst presenting a
+radical shift in function signature from other OpenAssetIO batch-first
+functions.
+
+```python
+class AbstractPage:
+
+    def get(self) -> List[List[EntityReferences]]:
+        # Current page of refs for each element of input batch.
+        ...
+
+    def next(self) -> AbstractPage:
+        ...
+
+    def remaining(self) -> Union[List[int], None]:
+        # None means "unsupported".
+        ...
+
+def relatedEntities(
+        batch: List[EntityReference],
+        relationship: TraitsData,
+        limit: int = -1) -> AbstractPage: ...
+```
+
+##### Pros
+
+- Retains batch-awareness for each page.
+- Retrieving the data is decoupled from traversal of pages, allowing for
+  lazy network retrieval of desired pages.
+- Additional metadata can be included in the object (e.g. `remaining`
+  pages) in an extensible way.
+- Provides good optimisation opportunities to the manager.
+
+##### Cons
+
+- Wasted processing of empty lists by the host when pages are exhausted
+  i.e. when there are still pages remaining for some inputs but not
+  others.
+- Additional metadata in the object is limited to whole-batch data.
+
+
+## Options considered
+
+### Option 1 - Whole-batch page token
+
+Host supplies a batch of entity references and a callback. The callback
+is called for each input entity, with one of its arguments being a
+page of related entities as a simple list.
+
+Once callbacks have been called, the function returns a page token. This
+can be supplied to a subsequent call to get the next page of results for
+all entities in the input batch simultaneously.
+
+Note that this could also be designed to use a cursor object with
+similar pros and cons, but that option was discarded (see Background
+section).
+
+```python
+class AbstractPageToken:
+    ...
+
+SuccessCallback = Callable[[int, List[EntityReferences]], None]
+
+def relatedEntities(
+        batch: List[EntityReference],
+        relationship: TraitsData,
+        successCb: SuccessCallback,
+        errorCb,
+        limit: int = -1,
+        pageToken: AbstractPageToken = None) -> AbstractPageToken: ...
+```
+
+#### Pros
+
+- Conceptually simple.
+- Whole batch is paged at once, retaining batch-awareness in the
+  manager for subsequent pages.
+- Per-element callback means it's trivial to skip processing elements of
+  the input batch whose pages are exhausted.
+- Matches the most common pattern of pageable HTTP APIs allowing e.g.
+  the page token to be a thin wrapper around a "real" page token(s).
+- Retains existing design in line with OpenAssetIO design principles,
+  i.e. non-paged signature is simply the same function but ignoring
+  optional arguments.
+- Manager plugin can remain conceptually stateless, since required state
+  is bundled up in the page token.
```

Comment:
[Question] Is this not true of the cursor based solutions as well? (More on this below at the cursor part)

Reply:
(Resolving in favour of the other thread)